www/app/blog/2024-11-05-pgvector-080.mdx

```diff
@@ -0,0 +1,193 @@
+    (3, 1, '[3, 1, -2]'),
+    (1, 2, '[1.1, 2.2, 3.3]'), 
+    (2, 2, '[1.1, 2.2, 3.3]'),
+    (3, 2, '[1.1, 2.2, 3.3]');
+CREATE INDEX ON filtest USING hnsw (embedding vector_cosine_ops);
+```
+
+My table has 6 rows in two categories. Note that all vectors in category 2 are very similar to `[1,2,3]`, 
+vectors in category 1 are quite different (in fact, they are almost **orthogonal** - meaning that they represent unrelated data).
+
+What would you expect the following query to return?
+
+```sql
+select id, category, embedding<=>'[1,2,3]'  AS distance 
+FROM filtest where category=1 
+order by distance limit 3;
+```
+
+I asked for 3 vectors, from category 1, that are closest to `[1,2,3]`, so the correct answer 
+is to return all vectors from category 1. However, this is what I expected pgvector 0.7.4 to do:
+
+1. Scan the vector index and find the 3 vectors closest to `[1,2,3]`, some of which should belong to category 2.
+2. Filter the results and keep only vectors from category 1.
+3. Return partial results.
+
+However, in practice:
+
+```sql
+SELECT extversion FROM pg_extension WHERE extname = 'vector';
+ extversion
+------------
+ 0.7.4
+(1 row)
+
+select id, category, embedding<=>'[1,2,3]'  AS distance 
+FROM filtest where category=1 order by distance limit 3;
+ id | category |      distance
+----+----------+--------------------
+  3 |        1 | 1.0714285714285714
+  1 |        1 | 1.0714285714285714
+  2 |        1 | 1.0714285714285714
+```
+
+How can this happen? pgvector's default parameter for the number of rows to scan in the vector index is 40.
+So the index scan in fact returned the entire table, and after filtering, we got the right result. 
+Thats the problem with small examples... if I tried with 100 vectors, it would have worked.
+
+Lets try with a tweaked parameter:
+
+```sql
+SET hnsw.ef_search = 3;
+SET
+
+select id, category, embedding<=>'[1,2,3]'  AS distance FROM filtest where category=1 order by distance limit 3;
+ id | category |      distance
+----+----------+--------------------
+  3 |        1 | 1.0714285714285714
+  2 |        1 | 1.0714285714285714
+(2 rows)
+```
+
+Ah, this is the problem result that I expected! Now lets see how pgvector 0.8.0 fixes it: 
+
+```sql
+SELECT extversion FROM pg_extension WHERE extname = 'vector';
+ extversion
+------------
+ 0.8.0
+
+SET hnsw.ef_search = 3;
+SET
+
+ select id, category, embedding<=>'[1,2,3]'  AS distance FROM filtest where category=1 order by distance limit 3;
+ id | category |      distance
+----+----------+--------------------
+  1 |        1 | 1.0714285714285714
+  2 |        1 | 1.0714285714285714
+  3 |        1 | 1.0714285714285714
+(3 rows)
+```
+
+Great! pgvector 0.8.0 delivers! Right? Yes, but not in the way you think:
+
+```sql
+show hnsw.iterative_scan;
+ hnsw.iterative_scan
+---------------------
+ off
+(1 row)
+```
+
+Oops! `hnsw.iterative_scan` is disabled. So how come we get all results? 
+Use of `explain plan` shows that the index wasn't even used:
```

Comment:
explain plan is an oracle thing right?

Reply:
LOL. I didn't touch Oracle since 2012 but here we are...